src/FSharpPlus.TypeLits/Matrix.fs

```diff
@@ -0,0 +1,632 @@
+namespace FSharpPlus.TypeLits
+
+open TypeLevelOperators
+open System.Runtime.CompilerServices
+
+#nowarn "0042" // retype
+
+module MatrixHelpers =
+  open System
+  let inline internal retype (x: 'T) : 'U = (# "" x: 'U #)
+
+  type NatToArguments =
+    static member inline Invoke (x: ^X, f: _ list -> _) =
+      let inline call_2 (a: ^a, b: ^b) = ((^a or ^b): (static member Ctor: _*_*_*_->_) a,b,[],f)
+      let inline call (a: 'a, b: 'b) = call_2 (a, b)
+      call (Unchecked.defaultof<NatToArguments>, x)
+
+    static member inline Ctor (_: ^C, Z, xs, f) = f xs
+    static member inline Ctor (c: ^C, S n: S< ^n >, xs, f) =
+      fun x -> ((^C or ^n): (static member Ctor: _*_*_*_->_)c,n,x::xs,f)
+
+  module Constraints =
+    /// Constrain 't to be a nested tuple of <'t1,'t2,'t3,'t4,'t5,'t6,'t7,'tr>
+    let inline whenNestedTuple (t: 't) = 
+      (^t: (member Item1: 't1) t), (^t: (member Item2: 't2) t), (^t: (member Item3: 't3) t), (^t: (member Item4: 't4) t), (^t: (member Item5: 't5) t), (^t: (member Item6: 't6) t), (^t: (member Item7: 't7) t), (^t: (member Rest: 'tr) t)
+
+  type CountTuple =
+    static member inline Invoke xs : 'n =
+      let inline call_2 (a: ^a, b: ^b) = ((^a or ^b) : (static member CountTuple: _*_ -> _) b, a)
+      let inline call (a: 'a, b: 'b) = call_2 (a, b)
+      call (Unchecked.defaultof<CountTuple>, xs)
+
+    static member inline CountTuple (t: 't, ct: ^CountTuple) =
+      let _,_,_,_,_,_,_,tr : _*_*_*_*_*_*_* ^TR = Constraints.whenNestedTuple t
+      ((^TR or ^CountTuple): (static member CountTuple: _*_->_) tr,ct)
+      |> S |> S |> S |> S |> S |> S |> S
+
+    static member inline CountTuple (_: Tuple<_>, _: CountTuple) = S Z
+    static member inline CountTuple ((_, _), _: CountTuple) = S Z |> S
+    static member inline CountTuple ((_, _, _), _: CountTuple) = S Z |> S |> S
+    static member inline CountTuple ((_, _, _, _), _: CountTuple) = S Z |> S |> S |> S
+    static member inline CountTuple ((_, _, _, _, _), _: CountTuple) = S Z |> S |> S |> S |> S
+    static member inline CountTuple ((_, _, _, _, _, _), _: CountTuple) = S Z |> S |> S |> S |> S |> S
+    static member inline CountTuple ((_, _, _, _, _, _, _), _: CountTuple) = S Z |> S |> S |> S |> S |> S |> S
+    
+  type TupleToList =
```

Comment:
I think functions like this could be moved to the Tuple.fs and be reused from there.
Maybe they can even go in a separate PR.

Reply:
I'm going to merge FSharpPlus.TypeLits.fsproj into FSharpPlus.fsproj, then it would be good to move them to Tuple.fs while doing that

Reply:
Nice, please go ahead.